src/MongoDB/ReadPreference.c

```diff
@@ -63,9 +63,14 @@ static bool php_phongo_readpreference_init_from_hash(php_phongo_readpreference_t
 	}
 
 	if ((tagSets = zend_hash_str_find(props, "tags", sizeof("tags") - 1))) {
+		ZVAL_DEREF(tagSets);
 		if (Z_TYPE_P(tagSets) == IS_ARRAY) {
 			bson_t* tags = bson_new();
 
+			/* Separate tagSets as php_phongo_read_preference_prep_tagsets may
+			 * modify these tags. */
+			SEPARATE_ZVAL_NOREF(tagSets);
```

Comment:
Keeping with the tradition of using `ZVAL_DEREF` and `SEPARATE_ZVAL_NOREF`. If it's more sensible to use `SEPARATE_ZVAL` here (and in other cases where we separate), I'd like to create a separate ticket for this.

Reply:
Does this address the following comment from PHPC-1700:

> Additionally, Manager may also leak if an exception occurs after preparing the readPreferenceTags key in the URI options array. That may be possible if merge_context_options or phongo_manager_init throws.

The fix here is only for `php_phongo_readpreference_init_from_hash`, which Manager doesn't use.

Reply:
Manager does not leak in that case. manager-ctor-read_prefernece-error-001.phpt already checks for this by attempting to create a manager by combining primary read preference with tags. That test did not leak when I ran it with memcheck.

The leak does not originate from throwing an exception after preparing tagsets, but from the missing zval separation in `__set_state`. This is also evidenced by the change necessary to fix `bug1698-001.phpt`, which started failing after the fix. This is because the expectation there is for the tags to be an object after calling `__set_state`, which shouldn't happen if we separated the zval.